### ISSUE TITLE
Allow optional D8 or MFD routing in SDR and NDR

### ISSUE THREAD

**Issue by richpsharp** [natcap/invest#148]
SDR and NDR use MFD routing. Some researchers prefer D8 and some aren't sure but like to compare results against the two methods. This PEP would add:

* a change to the ndr/sdr model interface to allow a routing type (d8/mfd)
* a change to the user interface to allow users to select one or the other
* a change to the output values to indicate if the model was run with d8 or mfd like `n_export_d8_[additional_suffix].tif`

**Comment by phargogh** [comment]
If we don't get to this before the 3.9 release, we'll bump this to 3.10.

**Comment by richpsharp** [comment]
FYI, in later discussion it wasn't so much that "natcap" wants D8  so much as it was an offhand in a meeting with a collaborator who didn't know what MFD was. And digging into that the real reason was that they didn't like MFD streams were multi pixels wide (even though they are more accurate). I've got a ticket to work on narrowing the MFD streams in response to this. My only thought is that if you do go through with this, that you make MFD the default and a warning that running in D8 will give inaccurate results that are non-linearly sensitive to error in the DEM?

**Comment by dcdenu4** [comment]
Is it reasonable to remove the 3.9 release tag from this issue? Is there an argument the 3.9 release should wait for this to be implemented?

**Comment by phargogh** [comment]
Yep!  Absolutely reasonable.  I've removed this issue from the 3.9 release milestone.

After some discussion on the side, Rich's suspicion about narrower streams was confirmed, though there may also still be some value in supporting D8 for the sake of comparison as it's still the most commonly-used routing algorithm.  Rich will prototype this in https://github.com/therealspring/inspring and we'll see what the hydrologists think about it.

**Comment by richpsharp** [comment]
So digging into this even more, turns out our "partners" were the World Bank and Conservation International. Neither of them wanted MFD because there's no hydrology literature published on it (aside from the InVEST model itself). So we rewrote NDR in `inspring` to go D8 only. It also has changes where it dropped the subsurface routing, allows for a custom load raster to be passed in that can be selected via certain landcover types, and it intentionally only operates on a specific watershed which makes it amenable to parallelization for global runs. [link to ndr_plus for reference](https://github.com/therealspring/inspring/blob/feature/ndr-plus-1/src/inspring/ndr_plus/ndr_plus.py#L354)

Swapping MFD and D8 in and out isn't as trivial as swapping function calls since they both have some custom flow path walking in Cython modules underneath. So it's not so straightforward to do both without a lot of extra Cython code, plus you have to be careful about all the routing. 

My recommendation for InVEST would be to leave it alone since the specific issue that caused this to be raised has been resolved and they couldn't use the InVEST NDR anyway since they wanted that custom load layer in it. If you were to do both there's the implementation complexity which can be handled of course, but it's also going to need some kind of scientific analysis of why and how the results are different (they are significantly different) and guidance about which one to use... also raises the question of which one is "right". Anyway, just raising all of that to note it's not a couple week kind of thing and the engagement with natcap science should be significant. Plus the motivation for needing it is resolved.

So I'm inclined to "wontfix" this issue, but lets chat tomorrow during our call and make sure that still makes sense.

**Comment by dcdenu4** [comment]
Thanks for taking the time to provide a summary on this Rich.

**Comment by emlys** [comment]
@phargogh has partially completed this

**Comment by emlys** [comment]
Closing as duplicate of #1440
